make assert_equal an example how to partial `torch.testing.assert_close`

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60254 tests for diagnostics in callable `msg` in `torch.testing.assert_close`
* #60163 update docstring examples of `torch.testing.assert_close`
* #60091 Improve error messages of `torch.testing.assert_close` in case of mismatching values
* #58919 split TestAsserts by functionality
* **#58918 make assert_equal an example how to partial `torch.testing.assert_close`**
* #58917 introduce TestingErrorMeta for internal use
* #58916 remove pytest.UsageError
* #58915 hide top-level test functions from pytest's traceback
* #58914 remove rouge raise in helper function

~Instead of a distinct `torch.testing.assert_close` and `torch.testing.assert_equal`, this makes `torch.testing.assert_equal` a special case of `torch.testing.assert_close` for `rtol=atol=0`. In this case the closeness definition `abs(actual - expected) <= atol + rtol * abs(expected)` boils down to `abs(actual - expected) <= 0`. Since `abs(x)` can never be `<0`, this is equivalent to `abs(a - b) == 0` and this again boils down to `a == b`.~

Following https://github.com/pytorch/pytorch/pull/58918#issuecomment-860642057 and some offline discussions, we opted to use `assert_equal` as an example how to `partial` it.

This makes maintaing the module a lot easier, because we don't need to keep two functions in sync.

Differential Revision: [D29259404](https://our.internmc.facebook.com/intern/diff/D29259404)

pmeier committed:
make assert_equal a special case of assert_close

Instead of a distinct `torch.testing.assert_close` and
`torch.testing.assert_equal`, this makes `torch.testing.assert_equal` a
special case of `torch.testing.assert_close` for `rtol=atol=0`. In this
case the closeness definition `abs(actual - expected) <= atol + rtol *
abs(expected)` boils down to `abs(actual - expected) <= 0`. Since
`abs(x)` can never be `<0`, this is equivalent to `abs(a - b) == 0` and
this again boils down to `a == b`.

This makes maintaing the module a lot easier, because we don't need to
keep two functions in sync.

[ghstack-poisoned]

pmeier commented:
Instead of defining `assert_equal` ourselves, we could also use it as a good example on how to `partial` from `torch.testing.assert_close`. Thoughts @mruberry?

mruberry commented:
Is there logic like 

https://github.com/pytorch/pytorch/blob/3607478ecd42ee93da1f05e4bd75fc1cfdaf6335/torch/testing/_core.py#L107

somewhere, where this will short-circuit to a bitwise equivalence test?

mruberry commented:
Lint failures are real

pmeier commented:
> Is there logic like
> 
> https://github.com/pytorch/pytorch/blob/3607478ecd42ee93da1f05e4bd75fc1cfdaf6335/torch/testing/_core.py#L107
> 
> somewhere, where this will short-circuit to a bitwise equivalence test?

Nope, in the current state we will always use `torch.isclose` for comparison, i.e. checking if `abs(actual - expected) == 0` holds. I'm not sure if there are implementation details that would make this not equivalent to `==`. 

We could also use an equal check if `rtol == 0 and atol == 0`, but this is means more implementation since we need to handle `equal_nan` separately. Plus the differences would need to be documented. So if there is no apparent use case, I would leave it as is.

pmeier committed:
Update on "make assert_equal a special case of assert_close"


Instead of a distinct `torch.testing.assert_close` and
`torch.testing.assert_equal`, this makes `torch.testing.assert_equal` a
special case of `torch.testing.assert_close` for `rtol=atol=0`. In this
case the closeness definition `abs(actual - expected) <= atol + rtol *
abs(expected)` boils down to `abs(actual - expected) <= 0`. Since
`abs(x)` can never be `<0`, this is equivalent to `abs(a - b) == 0` and
this again boils down to `a == b`.

This makes maintaing the module a lot easier, because we don't need to
keep two functions in sync.

[ghstack-poisoned]

pmeier committed:
Update on "make assert_equal an example how to partial `torch.testing.assert_close`"


~Instead of a distinct `torch.testing.assert_close` and `torch.testing.assert_equal`, this makes `torch.testing.assert_equal` a special case of `torch.testing.assert_close` for `rtol=atol=0`. In this case the closeness definition `abs(actual - expected) <= atol + rtol * abs(expected)` boils down to `abs(actual - expected) <= 0`. Since `abs(x)` can never be `<0`, this is equivalent to `abs(a - b) == 0` and this again boils down to `a == b`.~

Following https://github.com/pytorch/pytorch/pull/58918#issuecomment-860642057 and some offline discussions, we opted to use `assert_equal` as an example how to `partial` it.

This makes maintaing the module a lot easier, because we don't need to keep two functions in sync.

[ghstack-poisoned]

pmeier commented on torch/testing/_asserts.py:
```diff
@@ -796,6 +633,24 @@ def assert_close(
+        >>> import torch
+        >>> assert_equal = functools.partial(torch.testing.assert_close, rtol=0, atol=0)
+        >>> assert_equal(1e-9, 1e-10)
+        AssertionError: Tensors are not close!
+
+        Mismatched elements: 1 / 1 (100.0%)
+        Greatest absolute difference: 8.999999703829253e-10 at 0 (up to 0 allowed)
+        Greatest relative difference: 8.999999583666371 at 0 (up to 0 allowed)
```
This will make much more sense after #60091 from later in this stack is in. The error will read:

```
AssertionError: Scalars are not equal!

Absolute difference: 8.999999703829253e-10
Relative difference: 8.999999583666371
```

pmeier committed:
Update on "make assert_equal an example how to partial `torch.testing.assert_close`"


~Instead of a distinct `torch.testing.assert_close` and `torch.testing.assert_equal`, this makes `torch.testing.assert_equal` a special case of `torch.testing.assert_close` for `rtol=atol=0`. In this case the closeness definition `abs(actual - expected) <= atol + rtol * abs(expected)` boils down to `abs(actual - expected) <= 0`. Since `abs(x)` can never be `<0`, this is equivalent to `abs(a - b) == 0` and this again boils down to `a == b`.~

Following https://github.com/pytorch/pytorch/pull/58918#issuecomment-860642057 and some offline discussions, we opted to use `assert_equal` as an example how to `partial` it.

This makes maintaing the module a lot easier, because we don't need to keep two functions in sync.

[ghstack-poisoned]

mruberry reviewed (approved):
Cool

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29259404).
